### ISSUE TITLE
[ entropy, security ] Do we want access to  for PTG.2 in EDN?

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#5639]
This is a follow on that came up as part of a discussion around #5630.

Essentially entropy_src has a ENTROPY_DATA register who's implementation predates CSRNG and EDN.   This ENTROPY_DATA register as implemented has some implementation flaws (due to it's realization as a prim_packer_fifo), and currently the POR is to remove this register entirely.

The question however is whether to replace it with a new functionality that can pass the equivalent entropy through CSRNG and EDN.   The questions here are:
1. Do we need such a feature if we already can have PTG.3 functionality in EDN? (though at 1/3 the  bandwidth achievable by a raw PTG.2 port)
2. What would the interface look like in EDN?
3. Is it worth a possible schedule extension?

**Comment by moidx** [comment]
@zi-v, @felixmiller it would be good to get your thoughts on this from a PTG.2 characterization/testing perspective, even if the mode is not going to be exposed in production mode. 

I assume we are still planning to support `FW_OV_RD_DATA`, which may be sufficient for characterization/testing of the raw entropy bits.

**Comment by felixmiller** [comment]
I am hesitant here. What speaks for having access to PTG.2 output would be:
- Having a fallback option if there is an issue with the CSRNG
- If we plan to implement a SW based DRG.4 at some point


@martin-lueker Can you quickly refresh my memory on the "though at 1/3 the bandwidth achievable by a raw PTG.2 port" aspect? Is that b/c you assume 128 bit output to the EDN (from CSRNG) per 384 bit input to CSRNG. Shouldn't we able to crank the CSRNG twice per 384 bit input and still comply with PTG.3 requipments?

**Comment by martin-lueker** [comment]
Regardless though we have some very compelling arguments to just close this as "NO".  The reason being that it's another feature and it helps us with defining diversification values as discussed in #4603.  (TL;DR: If we prohibit direct consumption of entropy source outputs, then we don't need diversification values for `entropy_src`, as all values PTG.3 or DRG.3 will pass through CSRNG, and we can tweak the personalization string mechanism depending on whether we are in DEV or PROD mode to close #4603.)   If we allow PTG.2 then we have to make something custom to add diversification values in entropy_src, and we have no infrastructure for that.

Are there any objections to just closing this then, and only allowing PTG.3?

**Comment by imphil** [comment]
I might be misunderstanding the issue, but just to confirm: We feed EDN1 into OTBN's RND register, which is providing randomness to OTBN software for key generation and similar purposes.

@felixmiller Is PTG.3 randomness is fine for those use cases?

**Comment by mwbranstad** [comment]
Last topology I saw, OTBN was getting both PTG.2 entropy from EDN0, and PTG.3 entropy from EDN1, assuming that EDN1 does a reseed after every GEN request, whereas EDN0 does not. Please update this statement if incorrect.

**Comment by martin-lueker** [comment]
@mwbranstad: I think you mean DRG.3 (i.e., CSRNG outputs with long-living seeds) and PTG.3 (near constant reseeding)

**Comment by imphil** [comment]
Yes, the question (for @felixmiller) here is especially around EDN1/PTG.3, since we pass this randomness directly on to software. (The randomness from EDN0 goes into a LFSR and is used only for blinding/masking and some secure wipe functionality, so that's less of a concern.)

**Comment by mwbranstad** [comment]
thanks @martin-lueker was trying to get the terms correct also.

**Comment by moidx** [comment]
+@vrozic whom is also familiar with AIS31 requirements.

Spec under question [AIS 20 31](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Zertifizierung/Interpretationen/AIS_20_AIS_31_Evaluation_of_random_number_generators_e.pdf?__blob=publicationFile)

It seems that PTG.3 should fulfill all use cases for EDN1. @felixmiller is authoritative here. 

@martin-lueker for the CSRNG software instance, is the interpretation then that it will operate with PTG.2 seeds? 

The remaining opens from @felixmiller still stand (for the software instance at least):

> *  Having a fallback option if there is an issue with the CSRNG
> *  If we plan to implement a SW based DRG.4 at some point

**Comment by martin-lueker** [comment]
Yes, PTG.2 seeds will be available for CSRNG.

**Comment by felixmiller** [comment]
@imphil and all: Yes, from a quality of random numbers perspective PTG.3 is an enhancement of PTG.2. PTG.3 random numbers can be used for any cryptographic operations where PTG.2 input would be sufficient.

**Comment by cdgori** [comment]
I believe (?) that PTG.3 random numbers are produced at (much) slower rate, so if there are performance requirements for any of those use cases we would need to check them.  @martin-lueker / @mwbranstad maybe could confirm the relative rates.

**Comment by martin-lueker** [comment]
TL;DR: "Much slower" is relative, standard specific and fortunately SW programmable.

Gory details:
If you wanted to call the result a fully SP 800-90C compliant "NRBG" then we are talking 3x slower.  (128 bits out per 384 seed). If you were targeting some use case where the cert lab said that you could get away with less decimation and still call it PTG.3 under AIS31 then I imagine that you could do a 1:1 ration of entropy-in:entropy-out, and maybe some lab would call that okay (?).    The good news being here that EDN will let you pick this ratio.

**Comment by mwbranstad** [comment]
@martin-lueker was reviewing this issue for any action items to resolve. Unless I missed something, I did not see any. Can we close this issue?

**Comment by mwbranstad** [comment]
Since I moved to close this months ago, will be closing in 5 days unless there is an objection.

**Comment by mwbranstad** [comment]
Closing by EOW if no new updates.

**Comment by martin-lueker** [comment]
I have no objection to closing this issue.

**Comment by cdgori** [comment]
Since the PTG.3 is available at 1/3 the rate of PTG.2 I think that resolves my "concern" (which was more just noting that we needed to know what the use cases were if we were going to get close to any performance thresholds - seems like we have programmable workarounds for all of that now though.)

No objection to close here.

**Comment by mwbranstad** [comment]
Based on the above, closing now.
